libs/client/src/storage.ts

```diff
@@ -76,6 +74,161 @@ async function initiateUpload(
+    input: {
+      content_type: contentType,
+      file_name: filename,
+    },
+    config,
+  });
+}
+
+type MultipartObject = {
+  partNumber: number;
+  etag: string;
+};
+
+async function partUploadRetries(
+  uploadUrl: string,
+  chunk: Blob,
+  config: RequiredConfig,
+  cancelled: { current: boolean },
+  tries: number,
+): Promise<MultipartObject> {
+  if (cancelled.current) {
+    throw new Error("Part upload failed, upload cancelled");
+  }
+
+  if (tries === 0) {
+    throw new Error("Part upload failed, retries exhausted");
+  }
+
+  const { fetch, responseHandler } = config;
+
+  try {
+    const response = await fetch(uploadUrl, {
+      method: "PUT",
+      body: chunk,
+    });
+
+    return (await responseHandler(response)) as MultipartObject;
+  } catch (error) {
+    console.error("Part upload failed, retrying", uploadUrl, error);
+    return await partUploadRetries(
+      uploadUrl,
+      chunk,
+      config,
+      cancelled,
+      tries - 1,
+    );
+  }
+}
+
+async function partUpload(
+  uploadUrl: string,
+  chunk: Blob,
+  config: RequiredConfig,
+  parallelUploads: Sema,
+  cancelled: { current: boolean },
+) {
+  await parallelUploads.acquire();
+
+  try {
+    return await partUploadRetries(uploadUrl, chunk, config, cancelled, 3);
+  } finally {
+    parallelUploads.release();
+  }
+}
+
+async function multipartUpload(
+  file: Blob,
+  config: RequiredConfig,
+): Promise<string> {
+  const { fetch, responseHandler } = config;
+  const contentType = file.type || "application/octet-stream";
+  const { upload_url: uploadUrl, file_url: url } =
+    await initiateMultipartUpload(file, config, contentType);
+
+  // Break the file into 10MB chunks
+  const chunkSize = 10 * 1024 * 1024;
+  const chunks = Math.ceil(file.size / chunkSize);
+
+  const parsedUrl = new URL(uploadUrl);
+
+  // Max 5 parallel uploads
+  const limitedParallelUploads = new Sema(3);
+
+  const partPromises: Promise<MultipartObject>[] = [];
+
+  // To be able to cancel the upload if any of the parts fail
+  const cancelled = { current: false };
+  for (let i = 0; i < chunks; i++) {
+    const start = i * chunkSize;
+    const end = Math.min(start + chunkSize, file.size);
+
+    const chunk = file.slice(start, end);
+
+    const partNumber = i + 1;
+    // {uploadUrl}/{part_number}?uploadUrlParams=...
+    const partUploadUrl = `${parsedUrl.origin}${parsedUrl.pathname}/${partNumber}${parsedUrl.search}`;
+
+    partPromises.push(
+      partUpload(
+        partUploadUrl,
+        chunk,
+        config,
+        limitedParallelUploads,
+        cancelled,
+      ),
+    );
+  }
+
+  let responses: MultipartObject[];
+  try {
+    // Does this wait for all to finish even if an early one fails?
+    responses = await Promise.all(partPromises);
+  } catch (error) {
+    // If any of the parts fail, cancel other uploads
+    cancelled.current = true;
+
+    console.error("Multipart upload failed, aborting upload", error);
+    throw error;
```

Comment:
I think it's safe to remove console.error and either re-throw the error or throw a new one with the custom message + the cause